### PR TITLE
Added embeddings drift check + embeddings support in TextData

### PR DESCRIPTION
**Pull request by nirhutnik**
image

### CODE REVIEW

File: deepchecks/nlp/text_data.py [hunk 1]
```diff
@@ -89,13 +90,18 @@ class TextData:
     categorical_properties : t.Optional[t.List[str]] , default: None
         The names of the categorical properties columns. If None, categorical properties columns are automatically
         inferred. Only relevant if properties is not None.
+    embeddings : t.Optional[Union[pd.DataFrame, str]] , default: None
```
Comment (Nadav-Barak) [inline deepchecks/nlp/text_data.py:93]: I am against the 'auto' option, calculating is a heavy operation which can be buggy and as such should be executed after text data is created. Similar to properties calculation

File: deepchecks/nlp/text_data.py [hunk 2]
```diff
@@ -265,6 +281,35 @@ def n_samples(self) -> int:
         else:
             return 0
 
+    @property
+    def embeddings(self) -> pd.DataFrame:
+        """Return the metadata of for the dataset."""
+        return self._embeddings
+
+    def calculate_default_embeddings(self, model: str = 'miniLM', file_path: str = 'embeddings.csv'):
+        """Calculate the default properties of the dataset."""
```
Comment (Nadav-Barak) [inline deepchecks/nlp/text_data.py:290]: add detailed docstring

File: deepchecks/nlp/utils/text_embeddings.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+# Copyright (C) 2021-2023 Deepchecks (https://www.deepchecks.com)
+#
+# This file is part of Deepchecks.
+# Deepchecks is distributed under the terms of the GNU Affero General
+# Public License (version 3 or later).
+# You should have received a copy of the GNU Affero General Public License
+# along with Deepchecks.  If not, see <http://www.gnu.org/licenses/>.
+# ----------------------------------------------------------------------------
+#
+"""Utils module for calculating embeddings for text."""
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_embeddings.py:11]: There is a file called llm_utils, I think you can merge between them

File: tests/nlp/utils/test_embeddings.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+# ----------------------------------------------------------------------------
+#
+"""Test for the embeddings module"""
+
+import numpy as np
+from hamcrest import assert_that, equal_to
+from deepchecks.nlp.utils.text_embeddings import calculate_default_embeddings
+
+
+def test_simple_embeddings():
```
Comment (Nadav-Barak) [inline tests/nlp/utils/test_embeddings.py:18]: add a test for more use cases (Nones, weird chars, etc..)

File: deepchecks/nlp/text_data.py [hunk 5]
```diff
@@ -443,6 +488,19 @@ def label(self) -> TTextLabel:
...
+    @property
+    def label_for_display(self) -> TTextLabel:
+        """Return the label defined in the dataset.
+
+        Returns
+        -------
+        TTextLabel
+        """
+        if self.is_multi_label_classification():
+            return [np.argwhere(x == 1).flatten().tolist() for x in self.label]
```
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/nlp/text_data.py:500]: Also note that in extreme multilabel this will be huge. My need to use `format_list` function

File: deepchecks/nlp/utils/nlp_plot.py [hunk 6]
```diff
@@ -230,3 +232,21 @@ def get_text_outliers_graph(dist: Sequence, data: Sequence[str], lower_limit: fl
         hoverdistance=-1)
 
     return fig
+
+
+def two_datasets_scatter_plot(plot_title, plot_data, test_dataset, train_dataset, dataset_names):
+    """Plot a scatter plot of two datasets."""
```
Comment (noamzbr) [inline deepchecks/nlp/utils/nlp_plot.py:238]: What is plot data? Document for future selves

File: requirements/nlp-requirements.txt [hunk 7]
```diff
@@ -1,3 +1,4 @@
 seqeval>=1.0.0
 nltk>=3.4.0,<=3.6.7
-textblob>=0.17.1
\ No newline at end of file
+textblob>=0.17.1
+umap-learn
```
Comment (noamzbr) [inline requirements/nlp-requirements.txt:4]: heavy install or ok install?

File: deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+from deepchecks.nlp import Context, TrainTestCheck
+from deepchecks.nlp.checks.multivariate_embeddings_drift_utils import run_multivariable_drift_for_embeddings
+from deepchecks.utils.strings import format_number
+
+__all__ = ['TextEmbeddingsDrift']
+
+
+class TextEmbeddingsDrift(TrainTestCheck):
+    """
+    Calculate drift between the entire train and test datasets using a model trained to distinguish between them.
```
Comment (noamzbr) [inline deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py:22]:
```suggestion
    Calculate drift between the train and test datasets using a model trained to distinguish between them in embedding space.
```
If not more than one line...

File: deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    Parameters
+    ----------
+    sample_size : int , default: 10_000
+        Max number of rows to use from each dataset for the training and evaluation of the domain classifier.
+    random_state : int , default: 42
+        Random seed for the check.
+    test_size : float , default: 0.3
+        Fraction of the combined datasets to use for the evaluation of the domain classifier
+    min_meaningful_drift_score : float , default 0.05
+        Minimum drift score for displaying drift in check. Under that score, check will display "nothing found".
```
Comment (noamzbr) [inline deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py:39]: Think that in this check we should display in any case

File: deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+        DeepchecksValueError
+            If the object is not a Dataset or DataFrame instance
+        """
+        train_dataset = context.train
+        test_dataset = context.test
+
+        sample_size = min(self.sample_size, train_dataset.n_samples, test_dataset.n_samples)
+        headnote = """
+        <span>
+        The shown features are the features that are most important for the domain classifier - the
```
Comment (noamzbr) [inline deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py:102]: First line shouldn't be here

File: deepchecks/nlp/checks/multivariate_embeddings_drift_utils.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+from sklearn.model_selection import train_test_split
+from umap import UMAP
+
+from deepchecks.core.check_utils.multivariate_drift_utils import auc_to_drift_score, build_drift_plot
+from deepchecks.nlp import TextData
+from deepchecks.nlp.utils.nlp_plot import two_datasets_scatter_plot
+from deepchecks.utils.dataframes import floatify_dataframe
+from deepchecks.utils.plot import DEFAULT_DATASET_NAMES
+
+SAMPLES_FOR_REDUCTION_FIT = 1000
```
Comment (noamzbr) [inline deepchecks/nlp/checks/multivariate_embeddings_drift_utils.py:28]: comment explaining what that is?

File: deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+            random_state: int = 42,
+            test_size: float = 0.3,
+            min_meaningful_drift_score: float = 0.05,
+            dimension_reduction_method: str = 'auto',
+            num_samples_in_display: int = 500,
+            **kwargs
+    ):
+        super().__init__(**kwargs)
+
+        self.sample_size = sample_size
```
Comment (noamzbr) [inline deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py:68]: Follow up on @Nadav-Barak s comment - check if embeddings are calculated, and quit (error) if not.

File: deepchecks/nlp/checks/multivariate_embeddings_drift_utils.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    # create new dataset, with label denoting whether sample belongs to test dataset
+    domain_class_df = pd.concat([train_sample_df, test_sample_df]).reset_index(drop=True)
+    domain_class_labels = pd.Series([0] * len(train_sample_df) + [1] * len(test_sample_df))
+
+    # reduce dimensionality of embeddings if needed.
+    # skips if not required ('none') or if number of features is small enough (< 30) in 'auto' mode.
+    if not (dimension_reduction_method == 'none' or (
+            dimension_reduction_method == 'auto' and domain_class_df.shape[1] < 30)):
+
+        if (dimension_reduction_method == 'auto' and with_display) or dimension_reduction_method == 'umap':
```
Comment (noamzbr) [inline deepchecks/nlp/checks/multivariate_embeddings_drift_utils.py:53]: Would suggest for clarity to put all these conditions into variables before entering the if code.

File: deepchecks/nlp/checks/multivariate_embeddings_drift_utils.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+    """Calculate multivariable drift on embeddings."""
+    # sample train and test datasets equally
+    train_sample = train_dataset.sample(sample_size, random_state=random_state)
+    test_sample = test_dataset.sample(sample_size, random_state=random_state)
+
+    train_sample_df = train_sample.embeddings
+    test_sample_df = test_sample.embeddings
+
+    # create new dataset, with label denoting whether sample belongs to test dataset
+    domain_class_df = pd.concat([train_sample_df, test_sample_df]).reset_index(drop=True)
```
Comment (noamzbr) [inline deepchecks/nlp/checks/multivariate_embeddings_drift_utils.py:45]: why work internally with DF? I suspect it's slower.

File: deepchecks/nlp/checks/multivariate_embeddings_drift_utils.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+        new_embeddings_test = domain_class_df.iloc[len(train_sample_df):]
+        train_sample.set_embeddings(new_embeddings_train, verbose=False)
+        test_sample.set_embeddings(new_embeddings_test, verbose=False)
+
+    x_train, x_test, y_train, y_test = train_test_split(floatify_dataframe(domain_class_df), domain_class_labels,
+                                                        stratify=domain_class_labels, random_state=random_state,
+                                                        test_size=test_size)
+
+    # train a model to disguise between train and test samples
+    domain_classifier = GradientBoostingClassifier(max_depth=2, random_state=random_state)
```
Comment (noamzbr) [inline deepchecks/nlp/checks/multivariate_embeddings_drift_utils.py:73]: How did you determine depth 2? That's what we use in the tabular version?

File: deepchecks/nlp/text_data.py (hunk 1 above)
Comment (nirhutnik) [reply deepchecks/nlp/text_data.py:93]: Doesn't really exist, it's a mistake in the docstring. Fixed in the next PR of docs

File: deepchecks/nlp/utils/text_embeddings.py (hunk 3 above)
Comment (nirhutnik) [reply deepchecks/nlp/utils/text_embeddings.py:11]: I took the function out of there. I think this is more organized this way (we discussed this 2 months ago actually)

File: deepchecks/nlp/text_data.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/text_data.py:500]: Right now this is a small fix for the current issue, which just  returns [0 0 0 0 0 0 1 0 0 0 0 0]. I think this should be solved in another task.

File: deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py (hunk 8 above)
Comment (nirhutnik) [reply deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py:22]: It is though :( was cutting it very close. 
How about "Calculate drift between the train and test datasets using a model trained to distinguish between their embeddings."?

File: deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py (hunk 12 above)
Comment (nirhutnik) [reply deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py:68]: The .embeddings attribute throws an exception if called without embeddings, like metadata and properties. Why do anything different?

File: deepchecks/nlp/checks/multivariate_embeddings_drift_utils.py (hunk 14 above)
Comment (nirhutnik) [reply deepchecks/nlp/checks/multivariate_embeddings_drift_utils.py:45]: The embeddings are already in a DF to begin with actually.

File: deepchecks/nlp/checks/multivariate_embeddings_drift_utils.py (hunk 15 above)
Comment (nirhutnik) [reply deepchecks/nlp/checks/multivariate_embeddings_drift_utils.py:73]: Yes. I think we lowered it to save time on calculations... Though not sure we need it. Will check how it affects.

File: deepchecks/nlp/text_data.py [hunk 16]
```diff
@@ -89,13 +90,18 @@ class TextData:
     categorical_properties : t.Optional[t.List[str]] , default: None
         The names of the categorical properties columns. If None, categorical properties columns are automatically
         inferred. Only relevant if properties is not None.
+    embeddings : t.Optional[Union[pd.DataFrame, str]] , default: None
+        The text embeddings for the samples (1 embeddings per sample). If None, no embeddings are set. If 'auto', the
+        embeddings are calculated using the default embeddings. If a DataFrame is given, it must contain the embeddings
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/text_data.py]: This was an old docstring, fixed

File: deepchecks/nlp/text_data.py (hunk 1 above)
Comment (nirhutnik) [reply deepchecks/nlp/text_data.py:93]: Fixed now

File: deepchecks/nlp/text_data.py [hunk 17]
```diff
@@ -265,6 +281,35 @@ def n_samples(self) -> int:
...
+        embeddings : pd.DataFrame
+            Embeddings to set.
+        verbose : bool, default: True
+            Whether to print information about the process.
+        """
+        if self._embeddings is not None and verbose is True:
+            warnings.warn('Embeddings already exist, overwriting it', UserWarning)
+
+        if embeddings is not None:
+            validate_length_and_type(embeddings, 'Embeddings', len(self))
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/text_data.py:314]: Note it's a different function, I split the existing one so I will use only these specific functionalities

File: deepchecks/nlp/text_data.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/nlp/text_data.py:504]: Can you please open a task for it?

File: deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py (hunk 12 above)
Comment (noamzbr) [reply deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py:68]: didn't know

File: deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py (hunk 10 above)
Comment (nirhutnik) [reply deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py]: Done (whole thing was a left over copy+paste)

File: deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+from deepchecks.utils.strings import format_number
+
+__all__ = ['TextEmbeddingsDrift']
+
+
+class TextEmbeddingsDrift(TrainTestCheck):
+    """
+    Calculate drift between the entire train and test datasets using a model trained to distinguish between them.
+
+    Check fits a new model to distinguish between train and test datasets, called a Domain Classifier.
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py]: Done

File: deepchecks/nlp/utils/nlp_plot.py (hunk 6 above)
Comment (nirhutnik) [reply deepchecks/nlp/utils/nlp_plot.py]: Done

File: deepchecks/nlp/checks/multivariate_embeddings_drift_utils.py [hunk 19]
```diff
@@ -0,0 +1,120 @@
+# ----------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/checks/multivariate_embeddings_drift_utils.py]: Done

File: deepchecks/nlp/text_data.py (hunk 2 above)
Comment (nirhutnik) [reply deepchecks/nlp/text_data.py]: Done

File: deepchecks/nlp/checks/multivariate_embeddings_drift_utils.py (hunk 15 above)
Comment (nirhutnik) [reply deepchecks/nlp/checks/multivariate_embeddings_drift_utils.py]: Doesn't seem to make a significant change. makes sense - there are only 10 features left after reduction, and this is Gradient Boosting which is less sensitive to depth of tree. I think we can pin this as a future research but I don't think it should be change ATM

File: deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py (hunk 8 above)
Comment (noamzbr) [reply deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py:22]: 👍

File: deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py (hunk 9 above)
Comment (noamzbr) [reply deepchecks/nlp/checks/train_test_validation/text_embeddings_drift.py:40]: Some cool clusters potentially? Even if not belonging to train or test

**Review by noamzbr — APPROVED** [review]

**Review by noamzbr — APPROVED** [review]
